### ISSUE TITLE
False positive unreachable after type:ignore (also possibly invalid erorr code)

### ISSUE THREAD

**Issue by mortoray** [python/mypy#12785]
In the below program, mypy still considers the return statement unreachable, even though the error on the previous line is ignored. I would expect that by stating the first error can be ignored, I wouldn't need to mark statements after it as ignore[unreachable]. But it appears, that even though mypy doesn't emit the first error, it still continues to trust its assessment of the flow control.

```
# Simulates a wrongly typed function in a types library (Redis get command)
def wrong_type_get() -> str|None:
	return "abc"
	
def get_int() -> int:
	value = wrong_type_get()
	assert isinstance(value, int) # type:ignore[unreachable]
	return value
```

I can kind of understand why it happens, but I'm expecting MyPy to dismiss it's own knowledge when I state that it is wrong.  Perhaps there is another comment I can put to that effect, like a `# type:trustme:[not-unreachable]`?

----

As an aside, the "unreachable" code being ignored seems wrong. The error message is correct, but with a a weird code. Removing the type:ignore above the output is:

```
error: Subclass of "str" and "int" cannot exist: would have incompatible method signatures  [unreachable]
```

```
$ env/bin/mypy --version
mypy 0.950 (compiled: yes)
$ env/bin/python --version
Python 3.10.4
```

**Comment by AlexWaygood** [comment]
Looks like a `cast` works more effectively here:

```python
from typing import cast

# Simulates a wrongly typed function in a types library (Redis get command)
def wrong_type_get() -> str|None:
	return "abc"
	
def get_int() -> int:
	value = cast(int, wrong_type_get())
	return value
```

**Comment by mortoray** [comment]
Yes, that works. Though I'm leaving the assert in, for a sanity check, and casting simply to `object`.  The true function can return a variety of types.

**Comment by JelleZijlstra** [comment]
I think mypy is behaving correctly here. A type ignore just ignores the error, it doesn't have other effects on how mypy treats types.
